Add drivers for RTL8192EU

Mange opened the pull request:
This includes D-Link Wireless N300 Nano USB Adapter DWA-131 (Rev E1; USB ID: 2001:3319).

To be 100% honest now, most of this is cargo culted on my part. I copied the package for RTL8192**D**U and worked from that one. I found a new repo that hosted the source code for this chipset's driver and updated everything I could find.

I can't try this out currently as it tries to download the source code from `http://sources.openelec.tv/mirror/RTL8192EU/RTL8192EU-9e65859.tar.xz` and that URL obviously does not exist. Can I get it to compile from a local copy of the repo so I can verify this is working first?

Any suggestions on how to proceeed?

**TODO:**
- [ ] Pass review.
- [x] Make sure driver loads correctly and works for me.

---

Might I just say, your build pipeline is amazing. Great job! It's welcoming, easy and has great output. Everything's top-notch. I can see you spend time refining this, and I want to commend all of you for it.

chewitt commented:
Our dislike of non-kernel realtek drivers is well known so no guarantees this will be accepted, but if the sources are found in github you need to include a mkpkg script and squash the commits. There is prior art to copy in tools/mkpkg/

Mange commented:
I've added a `mkpkg_rtl8192eu` script.

Since the conventions everywhere else seems to be "packages in lowercase", I changed this driver to be all lowercase as well. The other `RTL*` drivers are in uppercase, but they also don't have any `mkpkg`-scripts.
I think the convention of the rest of the project will trump this driver collection, but I'll change back on request.

Mange commented:
I've gotten this to work when branching out from the `v5.0.8` tag (as I want to get it to work in my current setup; don't want to migrate my entire house to dev branch). I'm waiting for the compile to finish on the `master` branch now and I'll squash up my changes in here if it succeeds.

Mange commented:
There. The build now passes on `master` (6.0-devel).

I'm running this successfully on 5.0.8.

louisnichols commented:
Please forgive my ignorance in the ways of the github, but does "This branch is up-to-date with the base branch" mean this will be part of 6.0?

lrusak commented:
@louisnichols No, and this is not the place for discussion.

Mange commented:
I'd love for it to be part of 6.0, but I'm pretty sure we're not getting it
in there. Maybe 6.1?

tis 29 sep 2015 23:56 Lukas Rusak <notifications@github.com> skrev:

> @louisnichols <https://github.com/louisnichols> No, and this is not the
> place for discussion.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenELEC/OpenELEC.tv/pull/4277#issuecomment-144202588>
> .
>

chewitt commented:
Please UPPERCASE the driver package to have naming consistent with other realtek packages. Also test with current master branch which is now on 4.3 kernel. Also, since you are sourcing the driver from your own github repo - where did it come from?

Mange commented:
Thanks for getting back to me.

> Please UPPERCASE the driver package to have naming consistent with other realtek packages.

Sure, will do. Should the `mkpkg` script also be uppercase? I've done so right now, but in a different commit so I can discard it or keep it after I squash.

> Also test with current master branch which is now on 4.3 kernel.

I'm compiling right now. I'll get back to you after I've run the latest `master` build on my machine.

> Also, since you are sourcing the driver from your own github repo - where did it come from?

Yeah, that was a really bad journey. I've detailed it a bit in the repo itself on the commits:

*   Mange/rtl8192eu-linux-driver@1387cf623d54bc2caec533e72ee18ef3b6a1db29
      
      **The official RTL8192EU linux driver from D-Link Australia**
      Version information: 20140812_rtl8192EU_linux_v4.3.1.1_11320
      2014-08-12
      version 4.3.1.1_11320
      **Source:** [ftp://files.dlink.com.au/products/DWA-131/REV_E/Drivers/DWA-131_Linux_driver_v4.3.1.1.zip](ftp://files.dlink.com.au/products/DWA-131/REV_E/Drivers/DWA-131_Linux_driver_v4.3.1.1.zip)

      This version does not currently work on newer kernels, but it does contain USB ID 2001:3319, which a lot of other repos in GitHub does not.
*   Mange/rtl8192eu-linux-driver@416453cadc3a40036d625c6f798151b4b970f8d6
    
    **No more NULL dereference and kernel panic on current kernels**
    Found patch on the web: [http://users.telenet.be/x86_64/Patches/rtl8192eu-k3.13.patch](http://users.telenet.be/x86_64/Patches/rtl8192eu-k3.13.patch)
    
    Before patch, kernel panic due to a NULL dereference in `rtw_ndev_notifier_register`; this patch solves that problem by not using these methods on kernel versions after a certain number, so it leads me to believe that this patch is meant to make the driver compatible with newer kernel versions.

I realize now that I should've made all this a lot more explicit, both in the repo and by making a release tag or two.

**So in summary:** D-Link does not publish this latest version anywhere other than their Australia FTP server, as far as I could find it. It might have changed by now since there's been some time. I found more than one version floating around, but this one looked like the "latest" from diffing them with each other. Despite this, it was still not really compatible with newer kernels. I found a patch for this latest version that adds support for 4.x series of kernels, which I then added to the repo.

Let me know if you want more from me to trust this code. I'd be willing to sign stuff with my private key and you could then sign that with yours after verifying that the code published on the FTP site is actually what is in the repo. The SHA of the commit would then be a stable and safe point of reference.

---

This version (ie. this PR) has been working fine for me on the 5.0.8 release for a few months now.

Mange commented:
> I'm compiling right now. I'll get back to you after I've run the latest master build on my machine.

FYI, this version does not compile on 4.3 kernel. I found a rumour that D-Link has released an updated version, so I'm pursuing that route right now.

Mange commented:
Okay, I've rebased on top of `master`, adjusted everything until it compiled and I've also deployed this build to my Raspberry PI and everything works perfectly.



> I found a rumour that D-Link has released an updated version, [...]

This turned out to be a misunderstanding of me. Someone referenced a "October update", but they meant "The patch I made in october". Anyway, I was able to make the driver compatible with Kernel 4.3 by using three different patches from around the web. You can check the [Mange/rtl8192eu-linux-driver commits](/Mange/rtl8192eu-linux-driver/commits) for a list of them; you can also review them from here if you so wish.

---

Let me know if there's anything else you need from me.

lrusak commented:
:+1:  Thanks for your contribution. I have added the sources to the ftp.

Mange commented:
Thank *you* for maintaining the most awesome media appliance OS. :-)
